### ISSUE TITLE
Add option to composer update to prefer current versions

### ISSUE THREAD

**Issue by tgr** [composer/composer#6601]
After dependencies have changed (so that the current installed packages do not fulfill the current requirements anymore), I want to be able to make the required changes without any non-required changes.

E.g. I have a composer.json with
```
"require": {
    "foo/foo": "1.*",
    "bar/bar": "1.*"
}
```
and composer has, at some point in the past, installed version `foo/foo` 1.2 and `bar/bar` version 1.0 based on that. Since then, the world has moved on, and the current version of `foo/foo` is 1.5. I update the code and it now says
```
"require": {
    "foo/foo": "1.*",
    "bar/bar": "2.*",
    "boom/boom": "1.*"
}
```
I want to upgrade `bar/bar` and install `boom/boom` but I do not want to touch `foo/foo` as that is not needed to fulfull the requirements and I want to minimize the number of packages that are touched (to limit download time, or code review time for a project with a strict security policy, or to reduce the potential for new errors). Currently the only way to do this is `composer update --with-dependencies bar/bar boom/boom` which is awkward (I need to know which packages have non-compatible requirements; it is the package manager's job to figure that out) and not necessarily minimal (if `boom/boom` depends on `foo/foo: 1.*`, then `foo/foo` will get updated even though there is no need for it).

An especially common use case of this is projects using [composer-merge-plugin](https://github.com/wikimedia/composer-merge-plugin) for managing a framework with extensions/packages/bundles which have their own library dependencies: one of the extensions gets updated (or a new one added), and I want to be able to limit `composer update` to that extension without updating every single library used by the framework and the other installed extensions.

It would be great to have an option like `composer update --prefer-current` which would try to keep already-installed packages at their current versions unless that would conflict with another requirement. (Alternatively, `composer install` could do that when run in a directory which already has a composer.lock file but the it does not fulfill all requirements of composer.json. Anecdotally, that seems to be what most people expect `composer install` to do.)

**Comment by tgr** [comment]
cc @bd808

**Comment by tgr** [comment]
I can work on this if you feel this would be a useful addition and give me some pointers on where to look.

**Comment by alcohol** [comment]
`composer update bar/bar`.

**Comment by tgr** [comment]
As I said above, `composer update` will 1) do unnecessary updates (if used with `--with-dependencies`; otherwise it risks bringing the system into a broken state); 2) is cumbersome to use as I have to manually collect the names of all the libraries that need to be updated. It is not a solution to the problem described here.

**Comment by alcohol** [comment]
I'm sorry, but `composer update` is simply not the same as `composer update bar/bar`. So either there is some miscommunication taking place here, or you need to make sure you are more explicit in your examples.

**Comment by tgr** [comment]
I wrote in the original post
>  Currently the only way to do this is `composer update --with-dependencies bar/bar boom/boom` which is awkward (I need to know which packages have non-compatible requirements; it is the package manager's job to figure that out) and not necessarily minimal (if `boom/boom` depends on `foo/foo: 1.*`, then `foo/foo` will get updated even though there is no need for it).

Could you give me some pointer on making it more explicit? Would a real-life example help?

**Comment by alcohol** [comment]
I was replying to your latest comment, which stated:

> As I said above, `composer update` will 1) do unnecessary updates

**Comment by alcohol** [comment]
But I do not see how 

```
composer update --with-dependencies bar/bar
```

does unnecessary updates. It merely updates the specified package and anything it depends on.

**Comment by davidbarratt** [comment]
@alcohol 
from the description:
> if boom/boom depends on foo/foo: 1.*, then foo/foo will get updated even though there is no need for it

Meaning that 
```
composer update --with-dependencies boom/boom
```
would update the already installed `foo/foo` from `1.2` to `1.5` since `foo/foo` is a requirement of `boom/boom`. However, it does **not** need to be updated because `1.2` satisfies `boom/boom`'s requirement of `1.*`. This causes unnecessary updates that @tgr is trying to avoid (since in our case, updates to any library must be explicitly reviewed).

The only work-around I can think of is to do:
```
composer update --with-dependencies boom/boom --dry-run
```
and collect a list of what is going to be updated, then manually list everything that is missing:
```
composer update boom/boom hay/hay
```
which will prevent `foo/foo` from being updated.

But as @tgr said... that is super annoying.

**Comment by alcohol** [comment]
So, it sounds more like you want some kind of version pinning mechanic.

**Comment by davidbarratt** [comment]
@alcohol uhhh...yes. But I feel like that mechanic already exists in the lock file? Basically just need the CLI to honor a dep that is already locked rather than updating it. So maybe a `--no-locked` option or something?

**Comment by davidbarratt** [comment]
So `--no-locked` (or `--skip-locked` if you prefer) would skip anything in the lock file **unless** it is on the whitelist of packages to update OR it is not in the lock file.

**Comment by alcohol** [comment]
I did not think of it like that yet. I can see how the language/terms could be ambiguous here. For me, the lock file simply reflects the state of your dependencies at some point in time, and allows for consistent reinstalls of said state. While the lock file specifies specific versions, running composer update would still allow those versions to change within the constraint range defined in your composer.json. 

When I say pinning though, I see it more as a fixed version constraint. And now as I am writing this, I realize we technically already support this I guess, albeit not in a very convenient form factor. But practically speaking, you could list packages you do not want to be updated in your root composer.json, with simply their current specific version as the constraint.

You can then keep track of which packages could potentially be updated as follows:

```
composer outdated
// see composer outdated --help for all options
```

**Comment by davidbarratt** [comment]
@alcohol right. That would be another work-around, but yeah not very convenient. And I have seen a lot of projects do it this way, but imho, locking in `composer.json` kind of defeats the purpose of having the lock file. It also means the root package will have to list out every dependent package and keep that list up to date. :( which the lock file already does for us.

Ideally there would be a `prefer-locked` (or `prefer-current`) config in `composer.json` with a CLI option (so it would be just like `prefer-lowest` or `prefer-stable` in a way). I think most people would want composer to upgrade the package if it's _absolutely necessary_, but I would also be fine with it throwing an exception and listing the packages that would have to be updated (by adding it to the whitelist of packages to update), in which case the option would be `no-locked` rather than `prefer-locked`.

**Comment by alcohol** [comment]
I am not sure how we would implement this. It is a very specific requirement/edge-case.

What is wrong with adopting the following workflow:

```
// determine packages out of date
composer outdated --direct

// attempt to update only those (lets say foo/bar has available: 1.2.0 => 1.4.0)
composer update foo/bar

// if a package does not update to the version you expected
// because of a dependency you did not whitelist (suppose you got foo/bar 1.3.0)
composer why-not foo/bar 1.4.0

// re-run update with required dependency
composer update foo/bar other/dep
```

**Comment by davidbarratt** [comment]
You're right.. this isn't a problem.

Here are the tests I ran through, but could not make this happen. Composer _already_ respects the lock file by default.

But let's use a real life example.

Let's start a new project:
```
$ composer require doctrine/dbal --prefer-lowest
Using version ^2.6 for doctrine/dbal
./composer.json has been created
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 7 installs, 0 updates, 0 removals
  - Installing doctrine/lexer (v1.0): Downloading (100%)         
  - Installing doctrine/inflector (v1.0): Downloading (100%)         
  - Installing doctrine/collections (v1.0): Downloading (100%)         
  - Installing doctrine/cache (v1.0): Downloading (100%)         
  - Installing doctrine/annotations (v1.0): Downloading (100%)         
  - Installing doctrine/common (v2.7.1): Downloading (100%)         
  - Installing doctrine/dbal (v2.6.0): Downloading (100%)         
doctrine/dbal suggests installing symfony/console (For helpful console commands such as SQL execution and import of files.)
Writing lock file
Generating autoload files
```

We'll update some of the deps to prevent a version conflict:
```
$ composer update doctrine/cache doctrine/collections
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 0 installs, 2 update, 0 removals
  - Updating doctrine/cache (v1.0 => v1.7.1): Loading from cache
  - Updating doctrine/collections (v1.0 => v1.5.0): Loading from cache
Writing lock file
Generating autoload files
```

And at a later date, let's add a new package **without** updating `doctrine/annotations`:
```
$ composer require doctrine/orm:^2.5.12
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 6 installs, 0 updates, 0 removals
  - Installing psr/log (1.0.2): Loading from cache
  - Installing symfony/debug (v3.3.10): Loading from cache
  - Installing symfony/polyfill-mbstring (v1.6.0): Loading from cache
  - Installing symfony/console (v3.3.10): Loading from cache
  - Installing doctrine/instantiator (1.1.0): Loading from cache
  - Installing doctrine/orm (v2.5.12): Downloading (100%)         
symfony/console suggests installing symfony/event-dispatcher ()
symfony/console suggests installing symfony/filesystem ()
symfony/console suggests installing symfony/process ()
doctrine/orm suggests installing symfony/yaml (If you want to use YAML Metadata Mapping Driver)
Writing lock file
Generating autoload files
```

Now let's try it a different way... let's start over (cleaned out dir)
```
$ composer require doctrine/dbal --prefer-lowest
Using version ^2.6 for doctrine/dbal
./composer.json has been created
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 7 installs, 0 updates, 0 removals
  - Installing doctrine/lexer (v1.0): Loading from cache
  - Installing doctrine/inflector (v1.0): Loading from cache
  - Installing doctrine/collections (v1.0): Loading from cache
  - Installing doctrine/cache (v1.0): Loading from cache
  - Installing doctrine/annotations (v1.0): Loading from cache
  - Installing doctrine/common (v2.7.1): Loading from cache
  - Installing doctrine/dbal (v2.6.0): Loading from cache
doctrine/dbal suggests installing symfony/console (For helpful console commands such as SQL execution and import of files.)
Writing lock file
Generating autoload files
```
```
$ composer update doctrine/cache doctrine/collections
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 0 installs, 2 updates, 0 removals
  - Updating doctrine/cache (v1.0 => v1.7.1): Loading from cache
  - Updating doctrine/collections (v1.0 => v1.5.0): Loading from cache
Writing lock file
Generating autoload files
```

which gives us this `composer.json`
```
{
    "require": {
        "doctrine/dbal": "^2.6"
    }
}
```
now let's manually edit it:
```
{
    "require": {
        "doctrine/dbal": "^2.6",
        "doctrine/orm": "^2.5.11"
    }
}
```

And update the package...
```
$ composer update doctrine/orm
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 6 installs, 0 updates, 0 removals
  - Installing psr/log (1.0.2): Loading from cache
  - Installing symfony/debug (v3.3.10): Loading from cache
  - Installing symfony/polyfill-mbstring (v1.6.0): Loading from cache
  - Installing symfony/console (v3.3.10): Loading from cache
  - Installing doctrine/instantiator (1.1.0): Loading from cache
  - Installing doctrine/orm (v2.5.12): Loading from cache
symfony/console suggests installing symfony/event-dispatcher ()
symfony/console suggests installing symfony/filesystem ()
symfony/console suggests installing symfony/process ()
doctrine/orm suggests installing symfony/yaml (If you want to use YAML Metadata Mapping Driver)
Writing lock file
Generating autoload files
```

Which, either way, leaves us with what is out of date:
```
$ composer outdated
doctrine/annotations      v1.0    v1.5.0  Docblock Annotations Parser
doctrine/common           v2.7.1  v2.8.1  Common Library for Doctrine projects
doctrine/dbal             v2.6.0  v2.6.2  Database Abstraction Layer
doctrine/inflector        v1.0    v1.2.0  Common String Manipulations with regard to casing and singular/plural rules.
doctrine/lexer            v1.0    v1.0.1  Base library for a lexer that can be used in Top-Down, Recursive Descent Parsers.
```

but let's go one step further and try and update the direct packages that are outdated as @alcohol suggests:
```
$ composer update --root-reqs
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 0 installs, 1 update, 0 removals
  - Updating doctrine/dbal (v2.6.0 => v2.6.2): Loading from cache
Writing lock file
Generating autoload files
```

No need to update annotations:
```
$ composer outdated
doctrine/annotations      v1.0    v1.5.0  Docblock Annotations Parser
doctrine/common           v2.7.1  v2.8.1  Common Library for Doctrine projects
doctrine/inflector        v1.0    v1.2.0  Common String Manipulations with regard to casing and singular/plural rules.
doctrine/lexer            v1.0    v1.0.1  Base library for a lexer that can be used in Top-Down, Recursive Descent Parsers.
```

If it had thrown an error, we could have only updated that packages that threw errors as @alcohol recommends. I mean it would be _nice_ (imho) if it just did that for you automatically, but I would say that's a nice to have. It does seem a little tricky to explain/train other developers how to do this (would be much easier to give them a single command). But for the most part, using the CLI works exactly as intended (unless you get an error, in which you should update those packages).

So as a general rule for projects that want to update as few packages as possible... I would say: Always whitelist packages when requiring or updating.

Any other concerns @tgr?
